### PR TITLE
drivers: flash: merger Generic SPI Flash driver 

### PR DESCRIPTION
**Pull request by findlayfeng**
Merged #5456.
Update spi api.
Add the CS/gpio logic.
Independent of spi flash config.

### CODE REVIEW

**Comment by findlayfeng** [comment]
#7532

**Review by tbursztyka — APPROVED** [review]
Seems fine to me (remove the 'etc' in last patch's commit title).

Note that you don't have to close and create a new PR each time you make changes to your patches: a git push -f will just do, and all will be updated in existing PR.

**Comment by findlayfeng** [comment]
嗯，我尝试过`push -f`但是，不清楚是因为有三次commit 还是因为是一个空格改制表符的空提交，一直成功但没有被修改。正好提示有一个不兼容的合并，于是我就新拉取了一个分支做处理。
Below from Google Translate
Well, I tried `push -f` , and it's unclear whether there were three commit or because it was an empty commit of a space change tab, it was successful but not modified. Just in case there is an incompatible merge, so I took a new branch to do the processing.

**Comment by findlayfeng** [comment]
@nashif 
一个奇怪的错误，看起来不是我的问题
Below from Google Translate
A strange error that does not seem to be my problem

**Comment by findlayfeng** [comment]
@tbursztyka Made more changes in this patch

File: drivers/flash/Kconfig [hunk 1]
```diff
@@ -36,6 +36,39 @@ config FLASH_PAGE_LAYOUT
...
+	  Flash Discoverable Parameters (SFDP).
+
+config SPI_FLASH_SFDP_SPI_NAME
+	string
+	prompt "SPI controller device name"
+	depends on SPI_FLASH_SFDP
+	default ""
+
+config SPI_FLASH_SFDP_INIT_PRIORITY
+	int
```
Comment (tbursztyka) [inline drivers/flash/Kconfig]: no prompt for that one?

File: drivers/flash/Kconfig [hunk 2]
```diff
@@ -36,6 +36,39 @@ config FLASH_PAGE_LAYOUT
...
+	bool
+	prompt "Generic SPI Flash driver (SFDP)"
+	depends on SPI && FLASH
+	depends on !SPI_LEGACY_API
+	help
+	  This is a generic driver for SPI Flash chips supporting Serial
+	  Flash Discoverable Parameters (SFDP).
+
+config SPI_FLASH_SFDP_SPI_NAME
+	string
```
Comment (tbursztyka) [inline drivers/flash/Kconfig]: get rid of prompt and put the prompt message here.

File: drivers/flash/spi_flash_sfdp.c [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+	for (i = 0; i < nph; i++) {
+		size_t len;
+		void *ptr;
+
+		SYS_LOG_DBG("[%d] id %04x v%d.%d addr %06x len %d",
+				i, get_phdr_id(&phdr[i]),
+				phdr[i].major_ver, phdr[i].minor_ver,
+				get_phdr_addr(&phdr[i]),
+				phdr[i].length);
+		switch (get_phdr_id(&phdr[i])) {
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: why a switch if there is only 1 case?

File: drivers/flash/spi_flash_sfdp.c [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+
+	if (offset < 0 || offset+len >= data->flash_size) {
+		return -ENODEV;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+	wait_for_flash_idle(dev);
+	i = 0;
+	hdr[i++] = data->read_opcode;
+	if (data->addr_len == 4) {
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: add an empty line before

File: drivers/flash/spi_flash_sfdp.c [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: same

File: drivers/flash/spi_flash_sfdp.c [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+	hdr[i++] = offset >> 8;
+	hdr[i++] = offset;
+	r = spi_transceive(&data->config, txbufs, ARRAY_SIZE(txbufs),
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: same

File: drivers/flash/spi_flash_sfdp.c [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+	struct spi_buf txbufs[] = {
+		{ .buf = hdr, .len = 1 + data->addr_len },
+		{ .buf = NULL, .len = data->write_dummy },
+		{ .buf = (void *)ptr, .len = len },
+	};
+	int r, i;
+
+	SYS_LOG_DBG("write: @%x len %x, opcode %x (page size %d)",
+		offset, len,
+		data->write_opcode, data->page_size);
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: hum, you can put that on previous line

File: drivers/flash/spi_flash_sfdp.c [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+	wait_for_flash_idle(dev);
+
+	r = spi_flash_cmd(dev, CMD_WRITE_ENABLE);
+	if (r != 0) {
+		goto err;
+	}
+
+	i = 0;
+	hdr[i++] = data->write_opcode;
+	if (data->addr_len == 4) {
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: add an empty line before (basicall: if assigned value is not used in the if condition: break a line).

File: drivers/flash/spi_flash_sfdp.c [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: same here (break a line if it's not another '}' etc...)

**Review by tbursztyka — CHANGES_REQUESTED** [review]
you will need to rebase and update that against latest SPI API. Also, add support for CS GPIO (see existing code such as spi_flash_w25qxxdv.c)

**Comment by findlayfeng** [comment]
commit 88688ce  来自 Merged #5456，我只是让它变得可合并而已,我在 commit ff2e7ec 中做了自己的工作，包括gpio_cs，几乎重写了。
From google translate
Commit 88688ce from Merged #5456, I just made it mergeable, I did my work in commit ff2e7ec, including gpio_cs, which was almost rewritten.
@tbursztyka

File: drivers/flash/spi_flash_sfdp.h [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+#define CONFIG_SPI_FLASH_SMRP_ARRAY_SIZE 8
+#endif
+#ifndef CONFIG_SPI_FLASH_LAYOUTS_ARRAY_SIZE
+#define CONFIG_SPI_FLASH_LAYOUTS_ARRAY_SIZE 1
+#endif
+#ifndef CONFIG_SPI_FLASH_SMPT_SIZE
+#define CONFIG_SPI_FLASH_SMPT_SIZE 16
+#endif
+
+typedef u32_t dword_t;
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.h]: why this typedef? Seems unnecessary

File: drivers/flash/spi_flash_sfdp.c [hunk 11]
```diff
@@ -1,558 +1,993 @@
 /*
  * Copyright (c) 2017 Crypta Labs Ltd
+ * Copyright (c) 2018 Findlay Feng
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: Merge this specific modification of sfdp to your first commit.

And this actual commit becomes one for adding microchip table support then.

File: drivers/flash/spi_flash_sfdp.c [hunk 12]
```diff
@@ -1,558 +1,993 @@
...
+	if (!data->spi) {
 		return -EIO;
 	}
 
-	data->config = (struct spi_config) {
-		.dev = spi_dev,
-		.frequency = CONFIG_SPI_FLASH_SFDP_SPI_FREQ_0,
-		.operation = SPI_WORD_SET(8),
-	};
+#if defined(SPI_FLASH_GPIO_SPI_CS)
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: Why not a Kconfig option for this? So it would be CONFIG_ prefixed then.

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: drivers/flash/spi_flash_sfdp.c (hunk 11 above)
Comment (findlayfeng) [reply drivers/flash/spi_flash_sfdp.c]: I will change it

File: drivers/flash/spi_flash_sfdp.c (hunk 12 above)
Comment (findlayfeng) [reply drivers/flash/spi_flash_sfdp.c]: oh...
This is a lapse

**Comment by fabled** [comment]
Also overriding the configuration SPI freq does not seem right. This happens for reading SFDP table, and when microchip flash is detected.

**Comment by findlayfeng** [comment]
I only tried the standard spi, it is estimated that the dual spi is working, the quad spi still needs some extra work, I have left a lot of todo, and can improve it when needed.
Currently I don't have the right hardware to test.

**Comment by findlayfeng** [comment]
There is a definition of the read frequency that should be used in the SFDP document. This frequency is not necessarily the same as the one specified in the chip documentation.

**Comment by findlayfeng** [comment]
I use the nrf5x spi driver, the memory address of the configuration information does not change without modifying the actual configuration of the spi. . .

**Comment by fabled** [comment]
It is still normal that lower frequency is needed to be used. I think the configuration provided frequency should be honored. I rebased my spi sfdp driver on top of current head just to get everything I needed working - e.g. seems the microchip global unlock is not yet supported properly.

**Comment by findlayfeng** [comment]
page 3
```
4.4 Clock Rate
SFDP compliant devices must support 50 MHz operation for the Read SFDP command (instruction 5Ah).
Devices may support a wider frequency range, but a controller can always run SFDP cycles at 50 MHz or
less and get valid results. 
```

**Comment by galak** [comment]
We have a spi_nor driver in tree.  Please update any features like SFDP to that driver.  Closing this PR at this point.

**Pull request closed without merging**